pre_commit/languages/coursier.py

```diff
@@ -0,0 +1,65 @@
+import contextlib
+import os
+from typing import Sequence, Generator
+from typing import Tuple
+
+from pre_commit.envcontext import envcontext
+from pre_commit.envcontext import Var
+from pre_commit.envcontext import PatchesT
+from pre_commit.hook import Hook
+from pre_commit.languages import helpers
+from pre_commit.prefix import Prefix
+from pre_commit.util import clean_path_on_failure
+
+ENVIRONMENT_DIR = 'coursier'
+
+get_default_version = helpers.basic_get_default_version
+healthy = helpers.basic_healthy
+
+
+def install_environment(
+        prefix: Prefix,
+        version: str,
+        additional_dependencies: Sequence[str],
+) -> None:
+    helpers.assert_version_default('coursier', version)
+    helpers.assert_no_additional_deps('coursier', additional_dependencies)
+
+    envdir = prefix.path(helpers.environment_dir(ENVIRONMENT_DIR, version))
+    with clean_path_on_failure(envdir):
+        os.mkdir(envdir)
+        helpers.run_setup_cmd(
+            prefix,
+            (
+                'coursier',
+                'install',
+                '--default-channels=false',
+                '--channel=.pre-commit-channel',
+                'pre-commit-hook',
```

Comment:
(I'm unfamiliar with how the tool works so I'm asking more!)

does this require the entrypoint to always be `pre-commit-hook`?  this might break the expectations with `entry` in the pre-commit configuration

Reply:
Yeah, this is a bit of a weird thing. The easiest way to think about coursier in this context is that it's a bit like docker in the sense that it's fetching runnable artifacts from remote repositories (jars instead of containers), storing them locally (in a folder instead of in your local docker image registry) and then running them at a later time.

It looks like the docker integration is using the `prefix` as the unique tag as a way around the fact that the local docker registry is kind of shared system wide. In this case though coursier is placing the artifacts in these namespaced/prefixed pre-commit environments so a predictable artifact location is all we need.